# Step 4 - Model and evaluation


## A. Import thư viện

In [1]:
# 1. Import library
import lib.step1ml_utilis_summaryPre_processing as pre
import lib.step2ml_utils_feature_selection as fea
import lib.step3ml_utils_model_design_testing_regression as reg
import lib.step4ml_utils_model_design_testing_explainability as exp
import lib.step5ml_utils_model_design_testing_visualize_models as vis
import lib.step6ml_utils_model_design_testing_geospatial_analysis as geo
import lib.step9ml_Text_Kmeans_Clustering as clu
#import lib.step11ml_utilis_model_design_time_series as clu
import matplotlib
import matplotlib_inline
%matplotlib inline


import warnings
warnings.filterwarnings("ignore")

# B. Model

### 1.0. Upload dữ liệu


In [2]:
import pandas as pd
dir_file = "data/avocado_model.csv"
df = pd.read_csv(dir_file)
df.head()

Unnamed: 0    fea_date  fea_average_price  fea_total_Volume  fea_item_4046  \
0           1  2015-12-27               1.33          64236.62        1036.74   
1           2  2015-12-20               1.35          54876.98         674.28   
2           3  2015-12-13               0.93         118220.22         794.70   
3           4  2015-12-06               1.08          78992.15        1132.00   
4           5  2015-11-29               1.28          51039.60         941.48   

   fea_item_4225  fea_item_4770  fea_total_Bags  fea_small_bags  \
0       54454.85          48.16         8696.87         8603.62   
1       44638.81          58.33         9505.56         9408.07   
2      109149.67         130.50         8145.35         8042.21   
3       71976.41          72.58         5811.16         5677.40   
4       43838.39          75.78         6183.95         5986.26   

   fea_large_bags  ...  Item_4046  Item_4225  Item_4770  Total_Bags  \
0           93.25  ...    1036.74   54454.85      48.16     8696.87   
1           97.49  ...     674.28   44638.81      58.33     9505.56   
2          103.14  ...     794.70  109149.67     130.50     8145.35   
3          133.76  ...    1132.00   71976.41      72.58     5811.16   
4          197.69  ...     941.48   43838.39      75.78     6183.95   

   Small_Bags  Large_Bags  XLarge_Bags          type  year  region  
0     8603.62       93.25         0.00  conventional  2015  Albany  
1     9408.07       97.49         0.00  conventional  2015  Albany  
2     8042.21      103.14         0.00  conventional  2015  Albany  
3     5677.40      133.76         0.00  conventional  2015  Albany  
4     5986.26      197.69         0.00  conventional  2015  Albany  

[5 rows x 95 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18249 entries, 0 to 18248
Data columns (total 95 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Unnamed: 0                      18249 non-null  int64  
 1   fea_date                        18249 non-null  object 
 2   fea_average_price               18249 non-null  float64
 3   fea_total_Volume                18249 non-null  float64
 4   fea_item_4046                   18249 non-null  float64
 5   fea_item_4225                   18249 non-null  float64
 6   fea_item_4770                   18249 non-null  float64
 7   fea_total_Bags                  18249 non-null  float64
 8   fea_small_bags                  18249 non-null  float64
 9   fea_large_bags                  18249 non-null  float64
 10  fea_xlarge_bags                 18249 non-null  float64
 11  total_volumn_item               18249 non-null  float64
 12  fea_year_                       

### 1.1. Chọn lst_k_best

In [4]:
#lst_k_best_option1 = ['fea_type_organic','fea_total_Volume', 'fea_region_LosAngeles', 'fea_item_4046', 'fea_region_Houston','fea_region_PhoenixTucson', 'fea_region_Denver', 
#              'fea_region_DallasFtWorth','fea_region_CincinnatiDayton', 'fea_region_Roanoke', 'fea_region_SouthCentral', 'fea_region_Nashville', 
#              'fea_large_bags', 'fea_month', 'fea_item_4225', 'fea_item_4770', 'fea_total_Bags','fea_small_bags','total_volumn_item' ]
lst_k_best_option1 = ['fea_average_price','fea_type_organic', 'fea_large_bags', 'fea_item_4046', 'fea_year_', 'fea_item_4225', 'fea_id','fea_small_bags', 'fea_total_Bags', 'fea_item_4770', 'fea_month', 
    'total_volumn_item', 'fea_total_Volume','fea_xlarge_bags' ]

lst_k_best_option2_scaler = ['fea_average_price','fea_type_organic', 'fea_year_', 'fea_month', 'fea_id' ,'fea_total_Volume_scaler', 'fea_item_4046_scaler','fea_item_4225_scaler', 'fea_item_4770_scaler', 'fea_total_Bags_scaler','fea_small_bags_scaler', 
              'fea_large_bags_scaler','fea_xlarge_bags_scaler', 'total_volumn_item_scaler']



In [5]:
data_model_k_best_option1 = df[lst_k_best_option1]
data_model_k_best_option2_scaler = df[lst_k_best_option2_scaler]

In [6]:
data_model_k_best_option1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18249 entries, 0 to 18248
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   fea_average_price  18249 non-null  float64
 1   fea_type_organic   18249 non-null  int64  
 2   fea_large_bags     18249 non-null  float64
 3   fea_item_4046      18249 non-null  float64
 4   fea_year_          18249 non-null  int64  
 5   fea_item_4225      18249 non-null  float64
 6   fea_id             18249 non-null  int64  
 7   fea_small_bags     18249 non-null  float64
 8   fea_total_Bags     18249 non-null  float64
 9   fea_item_4770      18249 non-null  float64
 10  fea_month          18249 non-null  int64  
 11  total_volumn_item  18249 non-null  float64
 12  fea_total_Volume   18249 non-null  float64
 13  fea_xlarge_bags    18249 non-null  float64
dtypes: float64(10), int64(4)
memory usage: 1.9 MB


In [7]:
data_model_k_best_option2_scaler.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18249 entries, 0 to 18248
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   fea_average_price         18249 non-null  float64
 1   fea_type_organic          18249 non-null  int64  
 2   fea_year_                 18249 non-null  int64  
 3   fea_month                 18249 non-null  int64  
 4   fea_id                    18249 non-null  int64  
 5   fea_total_Volume_scaler   18249 non-null  float64
 6   fea_item_4046_scaler      18249 non-null  float64
 7   fea_item_4225_scaler      18249 non-null  float64
 8   fea_item_4770_scaler      18249 non-null  float64
 9   fea_total_Bags_scaler     18249 non-null  float64
 10  fea_small_bags_scaler     18249 non-null  float64
 11  fea_large_bags_scaler     18249 non-null  float64
 12  fea_xlarge_bags_scaler    18249 non-null  float64
 13  total_volumn_item_scaler  18249 non-null  float64
dtypes: flo

### 1.2. Tách 2 bộ dữ liệu type
- conventional
- organic

In [8]:
data_conventional = pre.filter_dataframe(df=df, feature_name='type', name_type='conventional')

close


In [9]:
data_organic = pre.filter_dataframe(df=df, feature_name='type', name_type='organic')

close


In [10]:
data_model_k_best_conventional = data_conventional[lst_k_best_option1]
data_model_k_best_conventional_scaler = data_conventional[lst_k_best_option2_scaler]

In [11]:
data_model_k_best_organic = data_organic[lst_k_best_option1]
data_model_k_best_organic_scaler = data_organic[lst_k_best_option2_scaler]

### 1.3. Hàm Kỹ thuật khác và biểu đồ
1. Hàm thực hiện cross valiation
2. Hàm thực hiện K-Folds
3. Hàm GridSearchCV
4. Hàm RandomSearch
5. Hàm Select Model


## B1. Chọn Model - check nmodel
- original
- scaler

In [12]:
#reg.check_model(df=data_model_k_best_conventional, y="fea_average_price", choose="select")

In [13]:
#reg.check_model(df=data_model_k_best_option2_scaler, y="fea_average_price", choose="select")

## B2. Chọn model - check model - theo type
- conventional
- organic

In [14]:
#reg.check_model(df=data_model_k_best_conventional, y="fea_average_price", choose="select")

In [15]:
#reg.check_model(df=data_model_k_best_organic, y="fea_average_price", choose="select")

## B2. Model scaler - Check model - scaler vs type
- conventional
- organic

In [16]:
#reg.check_model(df=data_model_k_best_conventional_scaler, y="fea_average_price", choose="select")

In [17]:
#reg.check_model(df=data_model_k_best_organic_scaler, y="fea_average_price", choose="select")

#### Nhận xét:
- Chung ta sẽ chọn model hình XGBRegressor hoặc RandomForestRegressor
- Có hiện tưởng overfitting
- Cần cải thiện thêm model

## B2. Model - check model - lazy

In [18]:
reg.check_model_final_lazy(df=data_model_k_best_conventional, y='fea_average_price', choose='model')

100%|██████████| 42/42 [00:31<00:00,  1.34it/s]


Adjusted R-Squared  R-Squared  RMSE  Time Taken
Model                                                                         
ExtraTreesRegressor                          0.88       0.88  0.09        2.73
RandomForestRegressor                        0.86       0.86  0.10        4.43
XGBRegressor                                 0.86       0.86  0.10        0.82
LGBMRegressor                                0.85       0.85  0.10        0.35
HistGradientBoostingRegressor                0.85       0.85  0.10        1.39
BaggingRegressor                             0.84       0.84  0.10        0.47
KNeighborsRegressor                          0.74       0.74  0.13        0.16
GradientBoostingRegressor                    0.72       0.73  0.14        1.90
DecisionTreeRegressor                        0.70       0.70  0.14        0.09
ExtraTreeRegressor                           0.68       0.68  0.15        0.05
MLPRegressor                                 0.64       0.64  0.16        3.10
SVR                                          0.60       0.60  0.17        1.38
NuSVR                                        0.60       0.60  0.17        2.09
AdaBoostRegressor                            0.54       0.54  0.18        0.82
BayesianRidge                                0.23       0.23  0.23        0.02
LassoCV                                      0.23       0.23  0.23        0.15
ElasticNetCV                                 0.23       0.23  0.23        0.14
LassoLarsCV                                  0.23       0.23  0.23        0.04
LarsCV                                       0.23       0.23  0.23        0.05
Ridge                                        0.23       0.23  0.23        0.02
Lars                                         0.23       0.23  0.23        0.02
RidgeCV                                      0.23       0.23  0.23        0.02
LassoLarsIC                                  0.23       0.23  0.23        0.02
SGDRegressor                                 0.23       0.23  0.23        0.02
TransformedTargetRegressor                   0.23       0.23  0.23        0.02
LinearRegression                             0.23       0.23  0.23        0.01
HuberRegressor                               0.22       0.23  0.23        0.10
LinearSVR                                    0.22       0.22  0.23        0.37
OrthogonalMatchingPursuitCV                  0.21       0.21  0.23        0.03
PoissonRegressor                             0.12       0.12  0.25        0.02
GammaRegressor                               0.11       0.12  0.25        0.03
TweedieRegressor                             0.11       0.11  0.25        0.02
GeneralizedLinearRegressor                   0.11       0.11  0.25        0.02
OrthogonalMatchingPursuit                    0.06       0.07  0.25        0.02
ElasticNet                                  -0.01      -0.00  0.26        0.02
DummyRegressor                              -0.01      -0.00  0.26        0.02
Lasso                                       -0.01      -0.00  0.26        0.02
LassoLars                                   -0.01      -0.00  0.26        0.02
RANSACRegressor                             -0.46      -0.45  0.32        0.10
PassiveAggressiveRegressor                  -1.45      -1.44  0.41        0.02
KernelRidge                                -19.24     -19.14  1.18        3.02
GaussianProcessRegressor                -21201.40  -21100.70 38.08        7.24

In [19]:
reg.check_model_final_lazy(df=data_model_k_best_conventional, y='fea_average_price', choose='predictions')

100%|██████████| 42/42 [00:29<00:00,  1.43it/s]


Adjusted R-Squared  R-Squared  RMSE  Time Taken
Model                                                                         
ExtraTreesRegressor                          0.88       0.88  0.09        2.15
RandomForestRegressor                        0.86       0.86  0.10        4.24
XGBRegressor                                 0.86       0.86  0.10        0.79
LGBMRegressor                                0.85       0.85  0.10        0.30
HistGradientBoostingRegressor                0.85       0.85  0.10        1.30
BaggingRegressor                             0.84       0.84  0.10        0.49
KNeighborsRegressor                          0.74       0.74  0.13        0.18
GradientBoostingRegressor                    0.72       0.73  0.14        1.76
DecisionTreeRegressor                        0.70       0.70  0.14        0.11
ExtraTreeRegressor                           0.68       0.68  0.15        0.03
MLPRegressor                                 0.64       0.64  0.16        3.10
SVR                                          0.60       0.60  0.17        1.49
NuSVR                                        0.60       0.60  0.17        1.94
AdaBoostRegressor                            0.54       0.54  0.18        0.77
BayesianRidge                                0.23       0.23  0.23        0.02
LassoCV                                      0.23       0.23  0.23        0.13
ElasticNetCV                                 0.23       0.23  0.23        0.14
LassoLarsCV                                  0.23       0.23  0.23        0.04
LarsCV                                       0.23       0.23  0.23        0.04
Ridge                                        0.23       0.23  0.23        0.02
Lars                                         0.23       0.23  0.23        0.02
RidgeCV                                      0.23       0.23  0.23        0.02
LassoLarsIC                                  0.23       0.23  0.23        0.02
SGDRegressor                                 0.23       0.23  0.23        0.03
TransformedTargetRegressor                   0.23       0.23  0.23        0.02
LinearRegression                             0.23       0.23  0.23        0.02
HuberRegressor                               0.22       0.23  0.23        0.09
LinearSVR                                    0.22       0.22  0.23        0.34
OrthogonalMatchingPursuitCV                  0.21       0.21  0.23        0.03
PoissonRegressor                             0.12       0.12  0.25        0.02
GammaRegressor                               0.11       0.12  0.25        0.02
TweedieRegressor                             0.11       0.11  0.25        0.02
GeneralizedLinearRegressor                   0.11       0.11  0.25        0.02
OrthogonalMatchingPursuit                    0.06       0.07  0.25        0.02
ElasticNet                                  -0.01      -0.00  0.26        0.02
DummyRegressor                              -0.01      -0.00  0.26        0.02
Lasso                                       -0.01      -0.00  0.26        0.01
LassoLars                                   -0.01      -0.00  0.26        0.02
RANSACRegressor                             -0.46      -0.45  0.32        0.10
PassiveAggressiveRegressor                  -1.45      -1.44  0.41        0.02
KernelRidge                                -19.24     -19.14  1.18        2.84
GaussianProcessRegressor                -21201.40  -21100.70 38.08        6.65

## Nhận xét:
- ExtraTreesRegressor: model và predictions đều thực hiện 88%
- RandomForestRegressor: model và predictions đều thực hiện 86%